### DISCUSSION TITLE
Non-pattern arguments to pattern synonyms

### DISCUSSION THREAD

**Discussion by anka-213** [ghc-proposals/ghc-proposals#677] [Ideas]
I would like some kind of syntax to specify that the first few arguments of a pattern synonym should be term arguments and not patterns.
The context where I first encountered this was when trying the new `RequiredTypeArguments` in 9.10 with pattern synonyms, which doesn't work at all, but then realizing that it would be kind of surprising if it did, since the result type depends on the type argument, so the type can't be reversible.

```haskell
pattern TypeIs :: forall a b -> Typeable a => (a ~~ b) => TypeRep b
pattern TypeIs a <- (eqTypeRep (typeRep @a) -> Just HRefl)
    where TypeIs = TypeRep
```
gives the error "a not in scope" and then if I fix that, "Pattern synonym ‘TypeIs’ has one argument but its type signature has 1 fewer arrows".

With `ScopedTypeVariables`, this works, but only if `a` is a type literal at use
```haskell
pattern TypeIs :: forall a b. Typeable a => (a ~~ b) => TypeRep b
pattern TypeIs <- (eqTypeRep (typeRep @a) -> Just HRefl)
    where TypeIs = TypeRep
```

So, the type signature of a pattern synonym consists of three parts which I call: _Preconditions_, _Match arguments_ and _Result type_. I haven't checked what the actual names GHC uses for these is.

```haskell
pattern Foo :: forall a b. Show a => Show b => Maybe b -> (a :~~: b, Maybe a)
--             \-- Preconditions --/ \--- Match args ---/ \-- Result type --/
pattern Foo x <- (HRefl, x)
```

The preconditions are arguments for both the pattern form and the constructor form, while the Match arguments are outputs for the pattern form and input for the constructor form. 

Any time one uses a visible forall, it forces it and everything that comes after it to become a part of the Result Type, instead of remaining a precondition as I expected.

What I would like is:
- Some way to explicitly specify what should be preconditions and what should be part of the match args/result type
- Allow term arguments and not just invisible foralls and constraints to be in preconditions

```haskell
pattern EqualTo :: forall a. Eq a => a -> § Bool -> § a
pattern EqualTo a x <- ((==) a -> x)
```
Here I used § to symbolize the separation, so `a` is input to the pattern, while `x` is output from the pattern.

The perhaps slightly confusing behavior I am expecting from this at use would be that based on the type signature, the first n arguments to the pattern synonym would be interpreted as terms instead of patterns. Though, this might also make parsing more difficult, so explicitly annotating the parameters themselves instead of or in addition to the type signature might be a better choice.

----

Perhaps this could also be two different proposal ideas. One is a possible solution to [ghc-issue#9671]https://gitlab.haskell.org/ghc/ghc/-/issues/9671), the other is more narrowly to treat visible forall the same way as invisible forall in pattern type signatures.

**Comment by goldfirere** [comment]
Yes, this would be good.

I agree that there are three things in a pattern synonym type. I'll relate the terms you wrote to ones used elsewhere:

* *Pre-conditions*: These are the inputs. Right now, GHC supports two flavors of pre-conditions: the *universal* type variables and the *required* constraints.
* *Match arguments*: These are the outputs. Right now, GHC supports three flavors of match argument: the *existential* type variables, the *provided* constraints, and the normal *arguments*.
* *Result type*: This is the type of the scrutinee (the thing being matched against).

You are asking for a third flavor of pre-condition, for term arguments.

I see two syntactic challenges here; one is easy and the other is hard. (There are no semantic challenges: this is straightforward to implement.)

**Challenge 1**: How to separate out pre-conditions from match arguments in the type? This is easy: put pre-conditions before a second `forall` or `=>`, just as today. That is, we have a general form of pattern signature type that looks like `forall universals. requireds => forall existentials. provideds => match arguments -> result`. The natural way to add term inputs is to make it `forall universals. requireds => term arguments -> forall existentials. provideds => match arguments -> result`. That is, if any types appear to the left of `->`s which itself appears to the left of a `forall` or `=>`, those are inputs.

**Challenge 2**: How to separate out pre-conditions from match arguments in terms? This is hard. If I have

```
pattern IsLessThan :: forall a. Num a => a -> forall. a   -- use empty "forall." just as a separator; `() =>` would also work
```

and then see

```
f x = ...
  where
    g (IsLessThan x) = ...
```

does the `x` in the definition of `g` define a new `x`, shadowing the outer one? or is it an occurrence of the already-bound `x` from above? Because `x` is an input to `IsLessThan`, we definitely want it to be an occurrence. But the only way to know this is by knowing the type of `IsLessThan`. Today, GHC resolutely maintains the fact that typing does not influence name resolution. This simplifies (greatly) the architecture of GHC itself, but also is a boon to programmers, who don't need to think about types in order to understand variable scoping. 

Unless we want to change this design point, we'll have to come up with some way to mark inputs. So we'd need something like `g (IsLessThen x ->)` or some other symbol that terminates the list of inputs. Or maybe we label outputs? But that wouldn't be backward compatible. I think when this idea has come up before, this syntactic challenge has been the main torpedo in the hull. Maybe you can guide us out, though? I do think everything else would fall into place once we figure this syntax piece out.

**Reply by Ericson2314** [reply]
And note @anka-213 that in the latest version of https://github.com/ghc-proposals/ghc-proposals/blob/master/proposals/0448-type-variable-scoping.rst#42proposed-change-specification, it is now agreed upon than we should be better about distinguishing inputs/usages outputs/binders than perhaps the status quo implementation does. So we really do need the new syntax.

**Comment by phadej** [comment]
There is some previous discussion in https://github.com/haskell/core-libraries-committee/issues/197
